### PR TITLE
Stop glob'ing on podman cp

### LINKED ISSUE #3829: podman cp dereferences symlink in host context after filepath.Glob(srcPath)
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

    podman cp mycontainer:/testfile /tmp/test

may resolve a symlink in the host context (e.g. the file _/tmp/hostdir/hello.txt_) and copy it into 
the directory _/tmp/test_

**Steps to reproduce the issue:**

```
erik@laptop:/tmp$ mkdir /tmp/hostdir
erik@laptop:/tmp$ echo hello > /tmp/hostdir/hello.txt
erik@laptop:/tmp$ cat /tmp/build.sh 
container=$(buildah from scratch)                                                                                                    
mnt=$(buildah mount $container)

touch "$mnt/file*"
ln -s "/file*" "$mnt/testfile"
ln -s "/tmp/hostdir/hello.txt" "$mnt/file_symlink_to_hostfile"

buildah umount $container
buildah commit $container glob1

erik@laptop:/tmp$ buildah unshare bash /tmp/build.sh && container=$(podman create localhost/glob1 dummycmd) 
442548187b2cb9fb3ee953a1987b643d50e7c42549b731aacc8a5ac9ebef083e
Getting image source signatures
Copying blob d3c1b073cbdd done
Copying config 6492f94e90 done
Writing manifest to image destination
Storing signatures
6492f94e909731586eecbd470ad2a9c378dc87fc2eb7e3d89712dfd4011170b4
erik@laptop:/tmp$ rm -rf /tmp/test && mkdir /tmp/test && podman cp $container:/testfile /tmp/test

```

**Describe the results you received:**

```
erik@laptop:/tmp$ ls -l /tmp/test
total 4
-rw-rw-r-- 1 erik erik 0 Aug 15 20:43 'file*'
-rw-rw-r-- 1 erik erik 6 Aug 15 20:34  hello.txt
erik@laptop:/tmp$ 
```

**Describe the results you expected:**

I would not have expected to see the files _/tmp/test/hello.txt_ and _/tmp/test/file*_

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
erik@laptop:/tmp$ podman --version
podman version 1.5.0
WARN[0000] unable to find /home/erik/.config/containers/registries.conf. some podman (image shortnames) commands may be limited 
erik@laptop:/tmp$
```

**Output of `podman info --debug`:**

```
erik@laptop:/tmp$ podman info --debug
debug:
  compiler: gc
  git commit: ""
  go version: go1.10.4
  podman version: 1.5.0
host:
  BuildahVersion: 1.10.1
  Conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.0, commit: unknown'
  Distribution:
    distribution: ubuntu
    version: "18.04"
  MemFree: 199711514624
  MemTotal: 201407889408
  OCIRuntime:
    package: Unknown
    path: /home/erik/crun/crun/static-build/crun
    version: |-
      crun 0.7
      spec: 1.0.0
      +SELINUX +CAP +SECCOMP +EBPF +YAJL
  SwapFree: 107374178304
  SwapTotal: 107374178304
  arch: amd64
  cpus: 32
  eventlogger: journald
  hostname: laptop
  kernel: 5.0.0-23-generic
  os: linux
  rootless: true
  uptime: 36h 26m 18.07s (Approximately 1.50 days)
registries:
  blocked: null
  insecure: null
  search: null
store:
  ConfigFile: /home/erik/.config/containers/storage.conf
  ContainerStore:
    number: 195
  GraphDriverName: overlay
  GraphOptions:
  - overlay.mount_program=/usr/bin/fuse-overlayfs
  GraphRoot: /home/erik/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 102
  RunRoot: /run/user/1000
  VolumePath: /home/erik/.local/share/containers/storage/volumes

erik@laptop:/tmp$ 
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

    erik@laptop:/tmp$ buildah --version
    buildah version 1.10.1 (image-spec 1.0.1, runtime-spec 1.0.1-dev)
    erik@laptop:/tmp$ cat /etc/issue
    Ubuntu 18.04.3 LTS \n \l

### PR DESCRIPTION
**Pull request by jwhonce**
* Do not support wildcards
* Restore pause=true to avoid race conditions.  User can still do --pause=false to avoid container pause.

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

File: cmd/podman/cp.go [hunk 1]
```diff
@@ -19,7 +19,6 @@ import (
 	"github.com/containers/storage/pkg/archive"
 	"github.com/containers/storage/pkg/chrootarchive"
 	"github.com/containers/storage/pkg/idtools"
-	securejoin "github.com/cyphar/filepath-securejoin"
```
Comment (TomSweeneyRedHat) [inline cmd/podman/cp.go]: Why are you dropping securejoin?  I thought that was the prefered way to handle directory joins especially if a symlink was involved?

File: cmd/podman/cp.go [hunk 2]
```diff
@@ -84,10 +83,14 @@ func copyBetweenHostAndContainer(runtime *libpod.Runtime, src string, dest strin
...
 		return errors.Errorf("invalid arguments %s, %s you must specify paths", src, dest)
 	}
+
 	ctr := srcCtr
 	isFromHostToCtr := ctr == nil
 	if isFromHostToCtr {
 		ctr = destCtr
+		destPath = anchorContainerPath(destPath)
+	} else {
+		srcPath = anchorContainerPath(srcPath)
```
Comment (mtrmac) [inline cmd/podman/cp.go]: Doesn’t this completely break the use of in-container relative paths (the code involving `ctr.WorkingDir` below)?

File: cmd/podman/cp.go [hunk 3]
```diff
@@ -190,53 +178,45 @@ func copyBetweenHostAndContainer(runtime *libpod.Runtime, src string, dest strin
...
-		err := copy(src, destPath, dest, idMappingOpts, &destOwner, extract, isFromHostToCtr)
-		if lastError != nil {
-			logrus.Error(lastError)
+	if srcPath, err = cleanPath(srcPath); err != nil {
+		return err
+	}
+
+	if destPath, err = cleanPath(destPath); err != nil {
+		return err
+	}
```
Comment (mtrmac) [inline cmd/podman/cp.go]: This can redirect the paths from inside the container to outside. How is the use of such paths prevented?

File: cmd/podman/cp.go [hunk 4]
```diff
@@ -271,25 +261,8 @@ func parsePath(runtime *libpod.Runtime, path string) (*libpod.Container, string)
...
-
-func copy(src, destPath, dest string, idMappingOpts storage.IDMappingOptions, chownOpts *idtools.IDPair, extract, isFromHostToCtr bool) error {
-	srcPath, err := filepath.EvalSymlinks(src)
-	if err != nil {
-		return errors.Wrapf(err, "error evaluating symlinks %q", srcPath)
-	}
-
-	srcPath, srcfi, err := getPathInfo(srcPath)
+func copy(srcPath, destPath, dest string, idMappingOpts storage.IDMappingOptions, chownOpts *idtools.IDPair, extract, isFromHostToCtr bool) error {
+	srcfi, err := os.Stat(srcPath)
```
Comment (mtrmac) [inline cmd/podman/cp.go]: This re-evaluates symlinks.

Sure, `cleanPath` has evaluated them once, so this should “only” matter if the container is running and racing against this code — but the `Pause` option is off by default.

- Why is it off by default? That makes it basically pointless.
- Why is it even an option? We shouldn’t have “break my system” options.

File: cmd/podman/cp.go (hunk 1 above)
Comment (jwhonce) [reply cmd/podman/cp.go]: @TomSweeneyRedHat I dropped all calls that had hidden evals of the symlinks as that caused the initial issue.  We need to eval once.

File: cmd/podman/cp.go (hunk 2 above)
Comment (jwhonce) [reply cmd/podman/cp.go]: @mtrmac I was following the contract with docker cli where cp is documented as anchored at the container's root.

File: cmd/podman/cp.go (hunk 4 above)
Comment (jwhonce) [reply cmd/podman/cp.go]: @mheon Can you answer @mtrmac please?

Comment (mheon) [reply cmd/podman/cp.go]: Let me dig up the old PR where that fix happened - originally, pause was optional, but default-on. We made the determination that this was a bad idea (copying a multi-gigabyte file would keep a container paused for a few minutes on a slower system, for example - and that seemed bad). I think the consensus was to disable it but leave the option available for those who want to be safe.

Comment (mheon) [reply cmd/podman/cp.go]: https://github.com/containers/libpod/pull/3214 merged with pause as default-true, but I know we changed that - still looking for where

Comment (mheon) [reply cmd/podman/cp.go]: https://github.com/containers/libpod/pull/3279

Fixed here, with in an unrelated PR, with no comments as to why. Past-me was a bad open source citizen.

I think @mtrmac is probably write about avoiding the re-evaluation here, given that we do not guarantee that containers are paused.

File: cmd/podman/cp.go (hunk 2 above)
Comment (mtrmac) [reply cmd/podman/cp.go]: OK, behaving the same makes sense; still, the code should be simplified.
- I don’t see any reason to do the “anchor” step up here when it could just as well be done in the following `if isFromHostToCtr` statement that deals with all other path remapping, and consumes the value.
- The `if isFromHostToCtr` case still has an `if filepath.IsAbs(destPath)` / `else`, with the two behaving differently (and the second one probably dead)
- The `if !isFromHostToCtr` case has an `if filepath.IsAbs(srcPath)` check with no `else`. Sure, that technically can’t happen right now, but it’s confusing and risky to leave around that way.

File: cmd/podman/cp.go (hunk 4 above)
Comment (mtrmac) [reply cmd/podman/cp.go]: Well, there is `os.Lstat`, but AFAICS `copyFileWithTar` is using `Stat` and `os.Open` again, so, just fixing this one line isn’t enough anyway.

File: cmd/podman/cp.go (hunk 1 above)
Comment (mtrmac) [reply cmd/podman/cp.go]: Sure, the re-evaluation is a problem, but that’s not enough of a justification to just discard all pre-existing protections (… without even mentioning that in the PR / commit description, even).

This PR currently replaces a vulnerability in a corner case with an even easier-to-hit vulnerability.

File: cmd/podman/cp.go (hunk 3 above)
Comment (mtrmac) [reply cmd/podman/cp.go]: … and why aren’t there any tests that fail if this exploit succeeds? Maintaining the security of the chroot is a critical feature of the `cp` functionality.

**Comment by rhatdan** [comment]
BTW, I think we should go back to default to pausing the container.  Then if someone wants to have better performance is willing to take on the risk, they can run it with the --pause=false flag.

**Comment by rhatdan** [comment]
One issue with the pause is this will fail on rootless containers not running with cgroups V2, but I still think it is a good default.

**Comment by mheon** [comment]
We can't break `cp` for rootless containers by default - need to verify it still works.
